source/extensions/load_balancing_policies/round_robin/config.h

```diff
@@ -57,16 +57,18 @@ class Factory : public Common::FactoryBase<RoundRobinLbProto, RoundRobinCreator>
 public:
   Factory() : FactoryBase("envoy.load_balancing_policies.round_robin") {}
 
-  Upstream::LoadBalancerConfigPtr loadConfig(Server::Configuration::ServerFactoryContext&,
-                                             const Protobuf::Message& config) override {
-    auto active_or_legacy = Common::ActiveOrLegacy<RoundRobinLbProto, ClusterProto>::get(&config);
-    ASSERT(active_or_legacy.hasLegacy() || active_or_legacy.hasActive());
+  absl::StatusOr<Upstream::LoadBalancerConfigPtr>
+  loadConfig(Server::Configuration::ServerFactoryContext&,
+             const Protobuf::Message& config) override {
+    ASSERT(dynamic_cast<const RoundRobinLbProto*>(&config) != nullptr);
+    const RoundRobinLbProto& typed_config = dynamic_cast<const RoundRobinLbProto&>(config);
+    // TODO(wbocode): to merge the legacy and typed config and related constructors into one.
```

Comment:
does this get cleaner when we do this TODO?  This feels like a lot of churn and I don't understand the benefits.  it's tagged as optimal but is it more efficient?

Reply:
The LB extension point is a litte different with other extension point (like http filter). The lb config loading and lb instance creation is to strictly split to two part: loadConfig & create.

At previous implementation:

1. The input proto message of `loadConfig` may be two different types: config proto of LB, or proto of Cluster. The `loadConfig` implementation need to handle two completely different message at same method (by dynamic cast). Weird and not match most cases of our other extension points.
2. The input typed lb config of `create` may be two different types: prod one and legacy one. The `create` implementation need to handle two different typed configs (also by dynamic cast) and use different constructor.

After these PR:

Now the loadConfig will strictly only handle the config proto of LB.

After all TODOs is completed:

Only single typed lb config implementation and constructor will be used for single LB type.

---

The benefits:
1. No dynamic cast and if branch checks everywhere. I don't think it will bring performance improvement. But it make code cleaner.
2. After all these constructors is unified, we then could also consider to use more efficient structure to stores these configuration options. (note, now the proto structure is used when we construct the LB).

Reply:
yeah I guess I don't see it as much cleaner?  I'm happy to through it over to someone else for a second opinion

Reply:
Different persons, different opinions. May be just because I am too sensitive to the dynamic cast based code branch. Introduce a third one is a good solution.